docs/graprof.1
.\" Manpage for graprof.
.\" Contact <andreas@grapentin.org> to suggest improvements or report errors.

.TH man 1 "01 November 2013" "0.7" "graprof man page"
.SH NAME
graprof \- a profilng and trace analysis tool
.SH SYNOPSIS
graprof [-CFgM?V] [-o <file>] [--call-graph] [--flat-profile] [--no-gui]
 [--memory-profile] [--output=<file>] [--help] [--usage]
 [--version] <binary file> <trace file>
.SH DESCRIPTION
Graprof is a profiling and trace analysis tool for C projects.

The development of graprof has been inspired by the profiling tools gprof and mtrace, and their inability to work together. Based on the most basic profiling methods the GNU toolchain has to offer, graprof aims to provide a low-impact easy to use, high-precision profiling method, as well as a sophisticated trace analysis tool which, sadly, is not yet implemented as of the current 0.7 release.

The information in this file, as well as some additional documentation, can be found on <http://graprof.grapentin.org>.

If you have found a bug, please do not hesitate to report it at graprof's bug tracker found on <http://github.com/oaken-source/graprof/issues>.

If you have any suggestions, on how graprof can be improved, there are various ways to get in touch with us. Either send an email to <graprof@grapentin.org>, or visit us in #graprof on freenode irc, and we will see to answering your requests.

If you want to support graprof, or its developers, have a look at <http://graprof.grapentin.org?page=support>, there are numerous ways to support graprof, from donating to just using it and spreading the word! we would appreciate it :)
.SH QUICK-START GUIDE
To instruct your application to write program traces, you have to add
 to CFLAGS:   -finstrument-functions -g
 to LDFLAGS:  -lgraprof

and recompile.

This instrumented binary should now be invoked with GRAPROF_OUT set to the filename you want the trace data to be written to, for example

 GRAPROF_OUT=gp.out ./your_application

To analyze the trace, invoke the graprof application, and pass the path to your binary, and the path to the trace file as arguments, for example:

 graprof -t gp.out -FCMg -o your_app.profile your_app
 
The -FCMg switch tells graprof to print the full analysis result to the specified profile file. 

Alternatively, you can merge these two steps together and run:

 graprof -FCMg -o your_app.profile [--] your_app [OPTIONS ...]

This will by default persist the trace information for reuse in a file called graprof.out, which can be changed by setting GRAPROF_OUT in the environment.
.SH OPTIONS
Graprof takes a couple of arguments to control the amount of profiling output generated.
.TP
.BR \-C ", " \-\-call\-graph
The \fB \-C \fR option instructs graprof to generate a call graph from the input trace file
.PP
.TP
.BR \-F ", " \-\-flat\-profile
The \fB -F \fR option instructs graprof to generate a flat profile from the input trace file. If \fB \-C \fR is also specified, the flat profile listing will include ids to reference functions in the call graph.
.PP
.TP
.BR \-M ", " \-\-memory\-profile
The \fB \-M \fR option instructs graprof to generate a memory profile from the input trace file, if allocation events are available.
.PP
.TP
.BR \-g ", " \-\-no\-gui
If the \fB \-g \fR option is given, graprof will not launch the trace visualization GUI, which is currently still in development. At this time, specifying this option results in an \fB ENOSYS \fR occurring, and a corresponding warning printed.
.PP
.TP
.B \-o \fIfilename\fR", "\fB\-\-output=\fIfilename\fR
This option allows specifying an output file, that the profiling results will be written out. Default is \fBstdout\fR
.PP
.TP
.B \-t \fIfilename\fR", "\fB\-\-trace=\fIfilename\fR
This option allows specifying a file as input that is used as trace instead of directly tracing the given binary. Set this option if you want to reuse older tracefiles generated from your application.
.PP
.TP
.BR \-v ", " \-\-verbose
Specify this option to increase the verbosity level of graprof. Specifying this flag multiple times will increase the verbosity even further. At verbosity level 1, the printed profile includes field descriptions.
.PP
.SH BUGS
Graprof is still mainly untested. There is only so much testing we can do, so this is mainly where graprof needs support from YOU! If you find any suspicious, reproducable behaviour, please file a bug on <http://github.com/oaken-source/graprof/issues>. If you want to send suggestions, regarding graprof itself, the website, or just want to chat with us a little, you can also mail <graprof@grapentin.org> or visit us in #graprof on freenode irc, and we will answer any queries you might have.

Apart from the lack of testing, graprof also has a couple of conceptual limitations, all of which are decribed in detail in the \fBREADME\fR included in the source distribution, and on the website <http://graprof.grapentin.org>
.SH SEE ALSO
gprof(1), mtrace(3)

Additional documentation can be found in the README of the source distribution and on the graprof website <http://graprof.grapentin.org>.
.SH COPYRIGHT
Copyright (C) 2013 - 2014  Andreas Grapentin <andreas@grapentin.org>

Permission is granted to copy, distribute and/or modify this document under the terms of the GNU Free Documentation License, Version 1.3 or any later version published by the Free Software Foundation; with no Invariant Sections, with no Front-Cover Texts, and with no Back-Cover Texts.  A copy of the license is included in the section entitled "GNU Free Documentation License".


NEWS
graprof NEWS -- history of user-visible changes. 2014-01-22
Copyright (C) 2013 - 2014 Andreas Grapentin
See the end for copying conditions

Please send graprof bug reports via 
<http://github.com/oaken-source/graprof/issues/>.

Version 0.7.1

* fixed a number of small dormant bugs

* improved portability and configure scripts


----------------------------------------------------------------------
Version 0.7

* Fixed crashes on concurrent access to allocation tracing functions; fixes #7

* altered profiling output for improved readability

* Notable performance boost in analysis 


----------------------------------------------------------------------
Version 0.6

* Changed invocation style to allow in-place execution of profiling child

* Notable performance increase in address translation and function comparison, 
  which is done A LOT.

* added trace file verification by comparing md5 sums of generating binary and
  binary passed to the analysis

* Fixed a bug where the analysis would crash if an address could be resolved
  to a function name but not a file or vice versa

* lots of minor tweaks and fixes


----------------------------------------------------------------------
Version 0.5

* This is the first beta release intended for public use. It requires the
  complete GNU toolchain, as well as the GNU C Library in a non-ancient
  version.

* Added a tracing library that hooks entry and exit calls to functions that
  are compiled with -finstrument-functions.

* Added runtime analysis comparable to that of gprof, with a couple of 
  differences that are subject to change.

* Added memory analysis comparable to that of mtrace, again with a couple
  of differences that are subject to change.


----------------------------------------------------------------------
Copyright information:

Copyright (C) 2013 - 2014 Andreas Grapentin

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.



graprof/graprof.c

/******************************************************************************
 *             graprof - a profiling and trace analysis tool                  *
 *                                                                            *
 *    Copyright (C) 2013 - 2014  Andreas Grapentin                            *
 *                                                                            *
 *    This program is free software: you can redistribute it and/or modify    *
 *    it under the terms of the GNU General Public License as published by    *
 *    the Free Software Foundation, either version 3 of the License, or       *
 *    (at your option) any later version.                                     *
 *                                                                            *
 *    This program is distributed in the hope that it will be useful,         *
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *    GNU General Public License for more details.                            *
 *                                                                            *
 *    You should have received a copy of the GNU General Public License       *
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.   *
 ******************************************************************************/


#include "graprof.h"

#include "addr.h"
#include "trace.h"
#include "flatprofile.h"
#include "callgraph.h"
#include "memprofile.h"

#include <grapes/feedback.h>
#include <grapes/file.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

FILE *graprof_out = NULL;

int
main (int argc, char *argv[])
{
  // parse arguments
  struct arguments arguments = { NULL, NULL, 0, NULL, 0 };
  argp_parse (&argp, argc, argv, ARGP_IN_ORDER, 0, &arguments);

  // set default trace file in environment, if not specified
  setenv("GRAPROF_OUT", "graprof.out", 0);

  // fork & exec the child
  if (!(arguments.trace_filename))
    {
      pid_t pid = fork();
      feedback_assert(pid != -1, "failed to fork `%s'", arguments.binary_invocation[0]);

      if (!pid) // child
        {
          execv(arguments.binary_invocation[0], arguments.binary_invocation);
          feedback_assert(0, "failed to execute `%s'", arguments.binary_invocation[0]);
        }
      else // parent
        {
          int status = 0;
          waitpid(pid, &status, 0);
          feedback_assert_wrn(WIFEXITED(status), "`%s': exited abnormally", arguments.binary_invocation[0]);
        }
    }

  // figure out the proper tracefile
  const char *tracefile = getenv("GRAPROF_OUT");
  if (arguments.trace_filename)
    tracefile = arguments.trace_filename;

  // read the tracefile
  int res = trace_read(tracefile);
  feedback_assert(!res, "`%s': corrupt trace data", tracefile);

  // determine the target output file
  graprof_out = stdout;
  if (arguments.out_filename)
    {
      graprof_out = fopen(arguments.out_filename, "w");
      feedback_assert(graprof_out, "`%s'", arguments.out_filename);
    }

  // determine what to do
  if (!arguments.tasks)
    arguments.tasks = GRAPROF_TRACING_GUI;

  // do it
  if (arguments.tasks & GRAPROF_FLAT_PROFILE)
    flatprofile_print(arguments.verbose);

  if (arguments.tasks & GRAPROF_CALL_GRAPH)
    callgraph_print(arguments.verbose);

  if (arguments.tasks & GRAPROF_MEMORY_PROFILE)
    memprofile_print();

  if (graprof_out != stdout)
    fclose(graprof_out);

  if (arguments.tasks & GRAPROF_TRACING_GUI)
    {
      #if 0
      res = traceview_main();
      feedback_assert(!res,
        (errno != ENOSYS
          ? "graprof traceview crashed"
          : "graprof traceview disabled at compile time"
        )
      );
      #endif
    }

  return 0;
}
